addons/stock/models/stock_move_line.py

```diff
@@ -448,6 +448,17 @@ def _action_done(self, cancel_backorder=False):
                         Quant._update_available_quantity(ml.product_id, ml.location_id, taken_from_untracked_qty, lot_id=ml.lot_id, package_id=ml.package_id, owner_id=ml.owner_id)
                 Quant._update_available_quantity(ml.product_id, ml.location_dest_id, quantity, lot_id=ml.lot_id, package_id=ml.result_package_id, owner_id=ml.owner_id, in_date=in_date)
             done_ml |= ml
+
+            # Unicity of serial tracked product. Search in quants if the sn is
+            # already present in stock
+            if ml.tracking == 'serial' and ml.lot_id:  # see test_inventory_4:
+                similar_quant = self.env['stock.quant'].search_count([
+                    ('lot_id', '=', ml.lot_id.id),
+                    ('quantity', '>', 0)
+                ])
+                if similar_quant > 1:
+                    raise ValidationError(_('You already have product %s with serial number %s.' % (ml.product_id.display_name, ml.lot_id.name)))
```

Comment:
UserError
don't like the error message but can't really think of better atm, maybe something like
"You cannot move product XX having a serial XX as it is already present elsewhere and the combination should be unique."